File: go/pkg/messaging/connector.go
package messaging

import (
	"fmt"

	"github.com/Bastien-Antigravity/microservice-toolbox/go/pkg/logger"
	"github.com/nats-io/nats.go"
)

// Connect establishes a connection to the NATS server using unified configuration options
// and routes NATS lifecycle events to the shared logger.
func Connect(cfg *NatsConfig, l logger.Logger) (*nats.Conn, error) {
	log := logger.EnsureSafeLogger(l)

	if len(cfg.Servers) == 0 {
		return nil, fmt.Errorf("no nats servers configured")
	}

	opts := []nats.Option{
		nats.Name(cfg.ClientID),
		nats.Timeout(cfg.ConnectTimeout),
		nats.ReconnectWait(cfg.ReconnectWait),
		nats.MaxReconnects(cfg.MaxReconnects),
		nats.FlusherTimeout(cfg.FlushTimeout),
		nats.RetryOnFailedConnect(true),

		// Standardized Connection Event Handlers
		nats.ClosedHandler(func(nc *nats.Conn) {
			log.Error("[%s] NATS connection closed unexpectedly", cfg.ClientID)
		}),
		nats.DisconnectErrHandler(func(nc *nats.Conn, err error) {
			log.Warning("[%s] NATS disconnected, attempting reconnect: %v", cfg.ClientID, err)
		}),
		nats.ReconnectHandler(func(nc *nats.Conn) {
			log.Info("[%s] NATS successfully reconnected to %s", cfg.ClientID, nc.ConnectedUrl())
		}),
	}

	conn, err := nats.Connect(cfg.Servers[0], opts...)
	if err != nil {
		return nil, fmt.Errorf("nats connection failed: %w", err)
	}

	log.Info("[%s] Successfully connected to NATS at %s", cfg.ClientID, conn.ConnectedUrl())
	return conn, nil
}

File: rust/src/messaging/connector.rs
use std::sync::Arc;
use std::time::Duration;
use crate::utils::logger::{Logger, ensure_safe_logger};
use crate::messaging::config::NatsConfig;

/// Establishes an asynchronous connection to the NATS server and configures event logging.
pub async fn connect(
    cfg: &NatsConfig,
    logger: Option<Arc<dyn Logger>>,
) -> Result<async_nats::Client, async_nats::Error> {
    let log = ensure_safe_logger(logger);

    if cfg.servers.is_empty() {
        return Err(async_nats::Error::from(std::io::Error::new(
            std::io::ErrorKind::InvalidInput,
            "no nats servers configured",
        )));
    }

    let log_disconnect = log.clone();
    let log_reconnect = log.clone();
    let client_id = cfg.client_id.clone();
    let client_id_reconnect = cfg.client_id.clone();

    let reconnect_wait_secs = cfg.reconnect_wait_secs;

    // Configure connection options with standard callbacks and properties
    let client = async_nats::ConnectOptions::new()
        .name(&cfg.client_id)
        .connection_timeout(Duration::from_secs(cfg.connect_timeout_secs))
        .reconnect_delay_callback(move |attempts| {
            log_disconnect.warning(&format!(
                "[{}] NATS disconnected, attempting reconnect (attempt {})...",
                client_id, attempts
            ));
            Duration::from_secs(reconnect_wait_secs)
        })
        .event_callback(move |event| {
            let log_reconnect = log_reconnect.clone();
            let client_id_reconnect = client_id_reconnect.clone();
            async move {
                if let async_nats::Event::Connected = event {
                    log_reconnect.info(&format!("[{}] NATS successfully reconnected", client_id_reconnect));
                }
            }
        })
        .connect(&cfg.servers[0])
        .await?;

    log.info(&format!("[{}] Successfully connected to NATS at {}", cfg.client_id, cfg.servers[0]));
    Ok(client)
}
